main/Main.hs

```diff
@@ -2,180 +2,290 @@
-        _ -> pure True
-    shouldRetry :: RetryStatus -> Bool -> IO Bool
-    shouldRetry rs isConnSucc = do
-      delay <- pure $ fromMaybe 0 (rsPreviousDelay rs) `div` 1000000
-      itShould <- pure $ not isConnSucc
-      when itShould $
-        putStrLn $ "Attempting to reconnect to the database in " <> (show delay::Text) <> " seconds..."
-      return itShould
+    -- The connection tries are capped, but if the connection times out no error is
+    -- thrown, just 'False' is returned.
+    connectingSucceeded :: IO Bool
+    connectingSucceeded =
+      retrying
+        (capDelay 32000000 $ exponentialBackoff 1000000)
+        shouldRetry
+        (const $ P.release pool >> isConnectionSuccessful)
+      where
+        isConnectionSuccessful :: IO Bool
+        isConnectionSuccessful = do
+          testConn <- P.use pool $ H.sql "SELECT 1"
+          case testConn of
+            Left e -> hPutStrLn stderr (toS $ encodeError e) >> pure False
+            _ -> pure True
+        shouldRetry :: RetryStatus -> Bool -> IO Bool
+        shouldRetry rs isConnSucc = do
+          delay <- pure $ fromMaybe 0 (rsPreviousDelay rs) `div` 1000000
+          itShould <- pure $ not isConnSucc
+          when itShould $
+            putStrLn $
+            "Attempting to reconnect to the database in " <>
+            (show delay :: Text) <>
+            " seconds..."
+          return itShould
+
+
+{-|
+  This is where everything starts.
 
+  Things that can throw errors: loadSecretFile
+                              , readOptions
+                              , when isMalformedProxyUri
+-}
 main :: IO ()
 main = do
+  --
+  -- LineBuffering: the entire output buffer is flushed whenever a newline is 
+  --                output
+  --              , the buffer overflows
+  --              , a hFlush is issued
+  --              , or the handle is closed
+  -- no-buffering: output is written immediately
+  --             , and never stored in the buffer
   hSetBuffering stdout LineBuffering
-  hSetBuffering stdin  LineBuffering
+  hSetBuffering stdin LineBuffering
   hSetBuffering stderr NoBuffering
-
+  --
+  -- readOptions builds the 'AppConfig' from the config file specified on the 
+  -- command line
   conf <- loadSecretFile =<< readOptions
   let host = configHost conf
       port = configPort conf
       proxy = configProxyUri conf
-      pgSettings = toS (configDatabase conf)
-      appSettings = setHost ((fromString . toS) host)
-                  . setPort port
-                  . setServerName (toS $ "postgrest/" <> prettyVersion)
-                  . setTimeout 3600
-                  $ defaultSettings
-
-  when (isMalformedProxyUri $ toS <$> proxy) $ panic
-    "Malformed proxy uri, a correct example: https://example.com:8443/basePath"
-
+      pgSettings = toS (configDatabase conf) -- is the db-uri
+      appSettings =
+        setHost ((fromString . toS) host) -- Warp settings
+         .
```

Comment:
This composition dot floating on its own line looks kind of odd. There's probably a more uniform way to align the composition symbol on each of these lines.